kratos/utilities/python_function_callback_utility.h

```diff
@@ -8,214 +8,357 @@
+        return mDependsOnSpace;
     }
 
-    double RotateAndCallFunction(const double x, const double y, const double z, const double t)
+    /**
+     * @brief This method returns the function body
+     * @return The function body
+     */
+    std::string FunctionBody()
+    {
+        return mFunctionBody;
+    }
+
+    /**
+     * @brief This method rotates and calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double RotateAndCallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        )
     {
         array_1d<double,3> xglobal;
         xglobal[0] = x;
         xglobal[1] = y;
         xglobal[2] = z;
-        array_1d<double,3> xlocal = prod(mR, (xglobal - mxc) );
-        return CallFunction(xlocal[0],xlocal[1],xlocal[2],t);
+        array_1d<double,3> xlocal = prod(mRotationMatrix, (xglobal - mCenterCoordinates) );
+        array_1d<double,3> xglobal_initial;
+        xglobal_initial[0] = X;
+        xglobal_initial[1] = Y;
+        xglobal_initial[2] = Z;
+        array_1d<double,3> xlocal_initial = prod(mRotationMatrix, (xglobal_initial - mCenterCoordinates) );
+        return CallFunction(xlocal[0],xlocal[1],xlocal[2], t,xlocal_initial[0],xlocal_initial[1],xlocal_initial[2]);
     }
 
-    double CallFunction(const double x, const double y, const double z, const double t)
+    /**
+     * @brief This calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double CallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        )
     {
-        main_namespace["x"] = x;
-        main_namespace["y"] = y;
-        main_namespace["z"] = z;
-        main_namespace["t"] = t;
-
-
-//         #if PY_MAJOR_VERSION >= 3
-//         PyObject* res = PyEval_EvalCode(mbytecode.ptr(),main_namespace.ptr(),main_namespace.ptr());
-//         #else
-//         PyObject* res = PyEval_EvalCode((PyCodeObject*)(mbytecode.ptr()),main_namespace.ptr(),main_namespace.ptr());
-//         #endif
-        return pybind11::eval(mfunction_body, main_namespace).cast<double>();
+        mMainNameSpace["x"] = x;
+        mMainNameSpace["y"] = y;
+        mMainNameSpace["z"] = z;
+        mMainNameSpace["X"] = X;
+        mMainNameSpace["Y"] = Y;
```

Comment:
the downaide of all of this is that we need to search for rhe parameters in rhe dict.
unfortunately i dont have any idea.of getting rid of this

Reply:
Me neither